File: multithreading/mat_mul_ex.py
#!/usr/bin/env python

#############################################
# Adapted from AM205 Fall 2020 by Nick Derr #
# https://github.com/nderr/multithreading   #
#############################################

# Python implementation of matrix multiplication using shared
# memory parallelization. Each takes two arguments:
# N, such that the system size of the multiplication is 2^N.
# t, the number of threads to parallelize over.

#################################
# 1. Imports and arg processing #
#################################

# Grab sys and timing modules
import sys
from time import time

# Pull user input
if len(sys.argv) < 2 or len(sys.argv) > 4:
    print("Usage: ./mat_mul.py [-v] N [t]")
    print("")
    print("  N: vector system is size 2^N (accept int,float)")
    print("  t: number of threads (optional, accept int)")
    print(" -v: use vectorized matrix multiply (optional)")
    sys.exit(0)

# Check if to use vectorized mat mul
vec=sys.argv[1]=='-v'

# Arg number for system size and threads
nArg=2 if vec else 1
tArg=3 if vec else 2

# Set number of threads if provided
userThreads=len(sys.argv)>tArg
if userThreads:
    nt=int(sys.argv[tArg])
else:
    nt=-1

# If vectorized, pass thread number through
# OMP environment variable
if nt>0 and vec:
    from os import environ
    environ['OMP_NUM_THREADS']=str(nt)
# OMP env change has to happen before imports

import joblib as jl
import numpy as np

########################
# 2. System allocation #
########################

# Set system size
N=int(2.**float(sys.argv[nArg]))

# Create random vector and matrix
v=np.random.uniform(size=(N,1),low=-1,high=1)
M=np.random.uniform(size=(N,N),low=-1,high=1)

#############################################
# 3. Helper function definitions for joblib #
#############################################

# Function for dotting a matrix row and the in-vector
def dot(r):
    #########################################
    ########### FILL IN CODE HERE ###########
    #########################################
    
    # Step 1: Grab reference to matrix and vector
    # Hint: Use "global" on two variables

    # global ?
    # global ?

    # Step 2: Loop over matrix columns to produce
    # requested value for each item in loop

    # total=0. # Create the dot product variable
    # then do a for loop

    # Step 3: Return dot product

    # return total

    #########################################
    ################## END ##################
    #########################################

# Function for parallel matrix multiplication
def mat_mul(nt):

    #########################################
    ########### FILL IN CODE HERE ###########
    #########################################
    
    # Step 1: Generator for components of dot product
    # Hint: Follow the example

    # fgen=(jl.delayed(?)(?) for ? in range(N))

    # Step 2: Parallel sum by forking rows

    # out=jl.Parallel(n_jobs=?)(?)

    # Step 3: Wrap in numpy object and return

    # return np.array(out)

    #########################################
    ################## END ##################
    #########################################

##############################################
# 4. Actual matrix multiplication and timing #
##############################################

# Minimum time for testing
Tmin=1
T=0

# Do mat muls until minimum time is reached
iters=0
t0=time()
while T<Tmin:

    # Case 1: Use joblib parallelized function
    if not vec:
        w=mat_mul(nt)
    # Case 2: Use numpy vectorized function
    else:
        #########################################
        ########### FILL IN CODE HERE ###########
        #########################################
        
        # Call numpy library function

        #########################################
        ################## END ##################
        #########################################

    # Record time and count iterate
    T=time()-t0
    iters=iters+1

# Print output
print("{:g}".format(T/iters))

File: multithreading/mat_mul_ex.cc
#include <cstdio>
#include <cstdlib>
#include <cmath>
#include <omp.h>

/*****************************************************************/
/*********** Adapted from AM205 Fall 2020 by Nick Derr ***********/
/************ https://github.com/nderr/multithreading ************/
/*****************************************************************/

/** C++ implementation of matrix multiplication using shared
 * memory parallelization. Each takes two arguments:
 * N, such that the system size of the multiplication is 2^N.
 * t, the number of threads to parallelize over. */

// Generate uniform random numbers between -1 and 1
inline double rnd() {
    return -1+2./RAND_MAX*static_cast<double>(rand());
}

/** Performs a matrix multiplication of a random matrix times
 * a random vector of length 2^N, using t threads.
 * \param[in] N log_2(system size)
 * \param[in] t number of OMP threads. */
int main(int argc,char** argv) {

    // Print a syntax message if a command-line argument isn't provided
    if(argc!=3) {
        puts("Usage: ./dot_product N [t]\n");
        puts("  N: vector system is size 2^N (accept int,float)");
        puts("  t: number of OMP threads (optional, accept int)");
        exit(0);
    }

    // Set OMP thread num
    int nt=1;
    // Only do this part if we've compiled with OpenMP
#if _OPENMP

    if (argc==3) {
        // If we have three args, user has specified nt
        nt=atoi(argv[2]);
    } else {
        // Otherwise, use default from OMP_NUM_THREADS
        nt=omp_get_max_threads();
    }
#endif

    // Minimum T for timing purposes
    static const double Tmin=0.5; // half second

    // Pull in vector size, allocate two vectors and a matrix
    int N=pow(2,atoi(argv[1]));
    double *v=new double[N];
    double *w=new double[N];
    double *M=new double[N*N];

    // Fill vectors/matrix with random values
#pragma omp parallel for num_threads(nt)
    for(int i=0;i<N;i++) {
        w[i]=0;
        v[i]=rnd();
        for(int j=0;j<N;j++) M[i+j*N]=rnd();
    }

    // Perform matrix multiplication
    double t0=omp_get_wtime(),T=0;
    int iters=0;

    // Do matrix multiplies for a speficied time
    while(T<Tmin) {

#pragma omp parallel for num_threads(nt)
        for(int i=0;i<N;i++) {
            for(int j=0;j<N;j++) {
                w[i]+=M[i+j*N]*v[j]
            }
        }
        
        T=omp_get_wtime()-t0; // Update time
        iters++; // Count iterations
    }

    // Print time per mat mult
    printf("%g\n",T/iters);

    // Cleanup vectors and matrix
    delete[] v;
    delete[] w;
    delete[] M;
}
